Source/Diagnostics/ReducedDiags/ColliderRelevant.cpp

```diff
@@ -0,0 +1,559 @@
+    if (!m_intervals.contains(step+1)) { return; }
+
+    // get MultiParticleContainer class object
+    const MultiParticleContainer& mypc = WarpX::GetInstance().GetPartContainer();
+
+    // get species names (std::vector<std::string>)
+    const std::vector<std::string> species_names = mypc.GetSpeciesNames();
+
+    // get a reference to WarpX instance
+    auto& warpx = WarpX::GetInstance();
+
+    // get cell size
+    amrex::Geometry const & geom = warpx.Geom(0);
+#if defined(WARPX_DIM_1D_Z)
+        amrex::Real dV = geom.CellSize(0);
+#elif defined(WARPX_DIM_XZ)
+        amrex::Real dV = geom.CellSize(0) * geom.CellSize(1);
+#elif defined(WARPX_DIM_3D)
+        amrex::Real dV = geom.CellSize(0) * geom.CellSize(1) * geom.CellSize(2);
+#endif
+
+    const auto get_idx = [&](const std::string& name){
+        return m_headers_indices.at(name).idx;
+    };
+
+    std::array<std::unique_ptr<amrex::MultiFab>, 2> num_dens;
+
+    // loop over species
+    for (int i_s = 0; i_s < 2; ++i_s)
+    {
+        // get WarpXParticleContainer class object
+        WarpXParticleContainer& myspc = mypc.GetParticleContainerFromName(m_beam_name[i_s]);
+        // get charge
+        amrex::ParticleReal const q = myspc.getCharge();
+
+        using PType = typename WarpXParticleContainer::SuperParticleType;
+
+        num_dens[i_s] = myspc.GetChargeDensity(0);
+        num_dens[i_s]->mult(1./q);
+
+        // wtot
+        amrex::Real wtot = ReduceSum( myspc,
+            [=] AMREX_GPU_HOST_DEVICE (const PType& p)
+            {
+                return p.rdata(PIdx::w);
+            });
+        amrex::ParallelDescriptor::ReduceRealSum(wtot, amrex::ParallelDescriptor::IOProcessorNumber());
+
+#if defined(WARPX_DIM_XZ)
+        // thetax_min, thetax_max
+        amrex::Real thetax_min = 0.0_rt;
+        amrex::Real thetax_max = 0.0_rt;
+
+        amrex::ReduceOps<ReduceOpMin, ReduceOpMax> reduce_ops_minmax;
+        auto r_minmax = amrex::ParticleReduce<amrex::ReduceData<Real, Real>>(
+            myspc,
+            [=] AMREX_GPU_DEVICE(const PType& p) noexcept -> amrex::GpuTuple<Real, Real>
+            {
+                const amrex::Real ux = p.rdata(PIdx::ux);
+                const amrex::Real uz = p.rdata(PIdx::uz);
+                const amrex::Real thetax = std::atan2(ux, uz);
+                return {thetax, thetax};
+            },
+            reduce_ops_minmax);
+        thetax_min = amrex::get<0>(r_minmax);
+        thetax_max = amrex::get<1>(r_minmax);
+        amrex::ParallelDescriptor::ReduceRealMin(thetax_min, amrex::ParallelDescriptor::IOProcessorNumber());
+        amrex::ParallelDescriptor::ReduceRealMax(thetax_max, amrex::ParallelDescriptor::IOProcessorNumber());
+
+        // x_ave, thetax_ave
+        amrex::Real x_ave = 0.0_rt;
+        amrex::Real thetax_ave = 0.0_rt;
+
+        amrex::ReduceOps<ReduceOpSum, ReduceOpSum> reduce_ops_ave;
+        auto r_ave = amrex::ParticleReduce<amrex::ReduceData<Real, Real>>(
```

Comment:
Could these ReduceOps be combined to reduce the number of kernels and avoid the duplication of the calculation of theta?